src/Compilers/CSharp/Portable/Symbols/VarianceSafety.cs

```diff
@@ -52,10 +54,70 @@ internal static void CheckInterfaceVarianceSafety(this NamedTypeSymbol interface
                     case SymbolKind.Event:
                         CheckEventVarianceSafety((EventSymbol)member, diagnostics);
                         break;
+                    case SymbolKind.NamedType:
+                        CheckNestedTypeVarianceSafety((NamedTypeSymbol)member, diagnostics);
+                        break;
                 }
             }
         }
 
+        /// <summary>
+        /// Check for illegal nesting into a variant interface.
+        /// </summary>
+        private static void CheckNestedTypeVarianceSafety(NamedTypeSymbol member, DiagnosticBag diagnostics)
+        {
+            switch (member.TypeKind)
+            {
+                case TypeKind.Class:
+                case TypeKind.Struct:
+                case TypeKind.Enum:
+                    break;
+                case TypeKind.Interface:
+                case TypeKind.Delegate:
+                    return;
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(member.TypeKind);
+            }
+
+            NamedTypeSymbol container = GetEnclosingVariantInterface(member);
+
+            if (container is object)
+            {
+                Debug.Assert(container.IsInterfaceType());
+                Debug.Assert(container.TypeParameters.Any(tp => tp.Variance != VarianceKind.None));
+                diagnostics.Add(ErrorCode.ERR_VarianceInterfaceNesting, member.Locations[0]);
+            }
+        }
+
+        internal static NamedTypeSymbol GetEnclosingVariantInterface(Symbol member)
```

Comment:
Should this return `NamedTypeSymbol?` since null is an intended part of the return type?

Reply:
> Should this return NamedTypeSymbol? since null is an intended part of the return type?

Nullable Reference Types feature hasn't been enabled in this file yet.


---
In reply to: [347049241](https://github.com/dotnet/roslyn/pull/39839#discussion_r347049241) [](ancestors = 347049241)